File: src/test/java/j/examples/validation/ValidationResultTest.java
package j.examples.validation;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

class ValidationResultTest {

    @Nested
    class AtomicValidationResults {
        @Test
        public void a_successful_result_is_Success() {
            assertThat(ValidationResult.success()).isSameAs(ValidationResult.Success.INSTANCE);
        }

        @Test
        public void a_failure_result_is_a_Failure_with_a_single_error_message() {
            var failure = ValidationResult.failure(MESSAGE);

            assertThat(failure).isInstanceOf(ValidationResult.Failure.class);
            assertThat(failure.errorMessages()).containsExactly(MESSAGE);
        }
    }

    @Nested
    class CombinedValidationResults {
        @Test
        public void an_empty_combination_yields_Success() {
            var combination = ValidationResult.combine();

            assertThat(combination).isInstanceOf(ValidationResult.Success.class);
        }

        @Test
        public void a_combination_of_a_single_Success_yields_Success() {
            var combination = ValidationResult.combine(ValidationResult.success());

            assertThat(combination).isInstanceOf(ValidationResult.Success.class);
        }

        @Test
        public void a_combination_of_a_single_Failure_yields_Failure() {
            var combination = ValidationResult.combine(ValidationResult.failure(MESSAGE));

            assertThat(combination).isInstanceOf(ValidationResult.Failure.class);
            assertThat(combination.errorMessages()).containsExactly(MESSAGE);
        }

        @Test
        public void a_combination_of_a_multiple_Successes_yields_Success() {
            var combination = ValidationResult.combine(
                ValidationResult.success(),
                ValidationResult.success()
            );

            assertThat(combination).isInstanceOf(ValidationResult.Success.class);
        }

        @Test
        public void a_combination_of_a_Success_and_Failure_yields_Success() {
            var combination = ValidationResult.combine(
                ValidationResult.success(),
                ValidationResult.failure(MESSAGE)
            );

            assertThat(combination).isInstanceOf(ValidationResult.Failure.class);
            assertThat(((ValidationResult.Failure) combination).errorMessages()).containsExactly(MESSAGE);
        }

        @Test
        public void a_combination_of_a_multiple_Failures_yields_Failure() {
            var combination = ValidationResult.combine(
                ValidationResult.failure(MESSAGE_1),
                ValidationResult.failure(MESSAGE_2)
            );

            assertThat(combination).isInstanceOf(ValidationResult.Failure.class);
            assertThat(combination.errorMessages()).containsExactly(MESSAGE_1, MESSAGE_2);
        }
    }

    @Nested
    class CombinedCollectionOfValidationResults {
        @Test
        public void a_combination_of_a_multiple_Failures_yields_Failure() {
            var combination = ValidationResult.combine(
                List.of(
                    ValidationResult.failure(MESSAGE_1),
                    ValidationResult.failure(MESSAGE_2)
                )
            );

            assertThat(combination).isInstanceOf(ValidationResult.Failure.class);
            assertThat(combination.errorMessages()).containsExactly(MESSAGE_1, MESSAGE_2);
        }
    }

    @Nested
    class GuardingBehavior {
        @Test
        public void checking_success_resumes_program_flow() {
            var result = ValidationResult.success();

            assertThatCode(() -> result.check()).doesNotThrowAnyException();
        }

        @Test
        public void checking_failure_aborts_program_flow_with_message() {
            var result = ValidationResult.failure(MESSAGE);

            assertThatThrownBy(() -> result.check())
                .isInstanceOf(ValidationException.class)
                .hasMessage(MESSAGE);
        }

        @Test
        public void checking_combined_failure_aborts_program_flow_with_all_messages() {
            var result = ValidationResult.combine(
                ValidationResult.failure(MESSAGE_1),
                ValidationResult.failure(MESSAGE_2)
            );

            assertThatThrownBy(() -> result.check())
                .isInstanceOf(ValidationException.class)
                .hasMessage("message-1,message-2");
        }
    }

    @Nested
    class ConditionBehavior {
        @Test
        public void evaluating_success() {
            var result = ValidationResult.success();

            assertThat(result.isSuccess()).isTrue();
        }

        @Test
        public void evaluating_failure() {
            var result = ValidationResult.failure("reason");

            assertThat(result.isSuccess()).isFalse();
        }
    }

    @Nested
    class FeedbackBehavior {
        @Test
        public void feedback_success_has_no_messages() {
            var result = ValidationResult.success();

            assertThat(result.errorMessages()).isEmpty();
        }

        @Test
        public void feedback_failure_has_a_message() {
            var result = ValidationResult.failure(MESSAGE);

            assertThat(result.errorMessages()).containsExactly(MESSAGE);
        }

        @Test
        public void feedback_combined_failure_has_multiple_messages() {
            var result = ValidationResult.combine(
                ValidationResult.failure(MESSAGE_1),
                ValidationResult.failure(MESSAGE_2)
            );

            assertThat(result.errorMessages()).containsExactly(MESSAGE_1, MESSAGE_2);
        }
    }

    private static final String MESSAGE = "message";
    private static final String MESSAGE_1 = "message-1";
    private static final String MESSAGE_2 = "message-2";
}

File: src/test/kotlin/k/examples/validation/ValidationResultTest.kt
package k.examples.validation

import k.examples.validation.ValidationResult.Failure
import k.examples.validation.ValidationResult.Success
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class ValidationResultTest {

    @Nested
    inner class `Atomic Validation Results` {
        @Test
        fun `A successful result is Success`() {
            assertThat(ValidationResult.success()).isSameAs(Success)
        }

        @Test
        fun `A failure result is a Failure with a single error message`() {
            val failure = ValidationResult.failure(MESSAGE)

            assertThat(failure).isInstanceOf(Failure::class.java)

            assertThat(failure.errorMessages).containsExactly(MESSAGE)
        }
    }

    @Nested
    inner class `Combined Validation Results` {
        @Test
        fun `An empty combination yields Success`() {
            val combination = ValidationResult.combine()

            assertThat(combination).isInstanceOf(Success::class.java)
        }

        @Test
        fun `A combination of a single Success yields Success`() {
            val combination = ValidationResult.combine(ValidationResult.success())

            assertThat(combination).isInstanceOf(Success::class.java)
        }

        @Test
        fun `A combination of a single Failure yields Failure`() {
            val combination = ValidationResult.combine(ValidationResult.failure(MESSAGE))

            assertThat(combination).isInstanceOf(Failure::class.java)
            assertThat((combination as Failure).errorMessages).containsExactly(MESSAGE)
        }

        @Test
        fun `A combination of a multiple Successes yields Success`() {
            val combination = ValidationResult.combine(
                ValidationResult.success(),
                ValidationResult.success(),
            )

            assertThat(combination).isInstanceOf(Success::class.java)
        }

        @Test
        fun `A combination of a Success and Failure yields Success`() {
            val combination = ValidationResult.combine(
                ValidationResult.success(),
                ValidationResult.failure(MESSAGE),
            )

            assertThat(combination).isInstanceOf(Failure::class.java)
            assertThat((combination as Failure).errorMessages).containsExactly(MESSAGE)
        }

        @Test
        fun `A combination of a multiple Failures yields Failure`() {
            val combination = ValidationResult.combine(
                ValidationResult.failure(MESSAGE_1),
                ValidationResult.failure(MESSAGE_2),
            )

            assertThat(combination).isInstanceOf(Failure::class.java)
            assertThat((combination as Failure).errorMessages).containsExactly(MESSAGE_1, MESSAGE_2)
        }
    }

    @Nested
    inner class `Combined Collection of Validation Results` {
        @Test
        fun `A combination of a multiple Failures yields Failure`() {
            val combination = ValidationResult.combine(
                listOf(
                    ValidationResult.failure(MESSAGE_1),
                    ValidationResult.failure(MESSAGE_2),
                )
            )

            assertThat(combination).isInstanceOf(Failure::class.java)
            assertThat((combination as Failure).errorMessages).containsExactly(MESSAGE_1, MESSAGE_2)
        }
    }


    @Nested
    internal class GuardingBehavior {
        @Test
        fun checking_success_resumes_program_flow() {
            val result = ValidationResult.success()
            Assertions.assertThatCode { result.check() }.doesNotThrowAnyException()
        }

        @Test
        fun checking_failure_aborts_program_flow_with_message() {
            val result = ValidationResult.failure(ValidationResultTest.MESSAGE)
            Assertions.assertThatThrownBy { result.check() }
                .isInstanceOf(ValidationException::class.java)
                .hasMessage(ValidationResultTest.MESSAGE)
        }

        @Test
        fun checking_combined_failure_aborts_program_flow_with_all_messages() {
            val result = ValidationResult.combine(
                ValidationResult.failure(ValidationResultTest.MESSAGE_1),
                ValidationResult.failure(ValidationResultTest.MESSAGE_2)
            )
            Assertions.assertThatThrownBy { result.check() }
                .isInstanceOf(ValidationException::class.java)
                .hasMessage("message-1,message-2")
        }
    }


    @Nested
    internal class ConditionBehavior {
        @Test
        fun evaluating_success() {
            val result = ValidationResult.success()
            assertThat(result.success).isTrue()
        }

        @Test
        fun evaluating_failure() {
            val result = ValidationResult.failure("reason")
            assertThat(result.success).isFalse()
        }
    }


    @Nested
    internal class FeedbackBehavior {
        @Test
        fun feedback_success_has_no_messages() {
            val result = ValidationResult.success()
            assertThat(result.errorMessages).isEmpty()
        }

        @Test
        fun feedback_failure_has_a_message() {
            val result = ValidationResult.failure(MESSAGE)
            assertThat(result.errorMessages).containsExactly(MESSAGE)
        }

        @Test
        fun feedback_combined_failure_has_multiple_messages() {
            val result = ValidationResult.combine(
                ValidationResult.failure(MESSAGE_1),
                ValidationResult.failure(MESSAGE_2)
            )
            assertThat(result.errorMessages).containsExactly(MESSAGE_1, MESSAGE_2)
        }
    }

    companion object {
        val MESSAGE = "message"
        val MESSAGE_1 = "message-1"
        val MESSAGE_2 = "message-2"
    }
}